src/components/admin/form-components/withSelector.jsx

```diff
@@ -0,0 +1,172 @@
+
+export const withSelector = Finder => {
+  class ObjectChip extends React.PureComponent {
+    onClick = () =>
+      this.props.onDelete(this.props.id, this.props.title, this.props.slug);
+
+    render() {
+      return (
+        <Chip onRequestDelete={this.onClick} style={this.props.style}>
+          {this.props.children}
+        </Chip>
+      );
+    }
+  }
+
+  ObjectChip.propTypes = {
+    onDelete: PropTypes.func.isRequired,
+    id: PropTypes.number,
+    title: PropTypes.string.isRequired,
+    slug: PropTypes.string,
+    style: PropTypes.shape({}),
+    children: PropTypes.node,
+  };
+
+  ObjectChip.defaultProps = {
+    id: null,
+    slug: null,
+    style: {},
+    children: null,
+  };
+
+  class Selector extends React.Component {
+    constructor(props) {
+      super(props);
+      this.handleDelete = this.handleDelete.bind(this);
+      this.handleClickAdd = this.handleClickAdd.bind(this);
+    }
+
+    componentDidUpdate(prevProps) {
+      if (!this.props.disabled && this.props.elements !== prevProps.elements) {
+        this.props.onChange();
+      }
+    }
+
+    handleClickAdd = async object => {
+      // disable this if saving
+      if (this.props.disabled) return;
+
+      const { id, title, slug } = object;
+      const alreadyAdded =
+        this.props.elements.find(
+          baseObject =>
+            !('slug' in baseObject)
+              ? this.props.slugify(baseObject.name) === slug
+              : baseObject.slug === slug,
+        ) !== undefined;
+
+      if (alreadyAdded) {
+        await this.props.displayAlert(
+          `The ${this.props.mode} ${title} has already been added.`,
+        );
+        return;
+      }
+      const newObjects = update(this.props.elements, {
+        $push: [{ id, name: title, slug }],
+      });
+      this.props.onUpdate(newObjects);
+    };
+
+    handleDelete(id, title, slug) {
+      // disabled this if saving
+      if (this.props.disabled) return;
+
+      const objectSlug = !slug ? this.props.slugify(title) : slug;
+
+      const index = this.props.elements.findIndex(
+        baseObject =>
+          !('slug' in baseObject)
+            ? this.props.slugify(baseObject.name) === objectSlug
+            : baseObject.slug === objectSlug,
+      );
+      if (index === -1) {
+        throw new Error(
+          'You tried to delete an object not currently selected. ' +
+            "This shouldn't happen, please let the developers know that it did.",
+        );
+      }
+      const newObjects = update(this.props.elements, { $splice: [[index, 1]] });
+      this.props.onUpdate(newObjects);
+    }
+
+    render() {
+      const styles = {
+        wrapper: {
+          display: 'flex',
+          flexWrap: 'wrap',
+        },
+      };
+
+      const objectChips =
+        this.props.elements.length > 0
+          ? _.map(this.props.elements, object => (
+              <ObjectChip
+                key={_.get(object, 'id', object.name)}
```

Comment:
This will return null when it's null, which is also bad sorry haha, if `object.name` is required you could just always use that as the name though right? Or are we afraid of collisions? Otherwise I think the easiest way is `object.id || object.id === 0 ? object.id : object.name` no?

Reply:
(actually, ids are always larger than 0 so you don't need the `||` at all)

Reply:
You are right, yes. Pushed it :)